### PR TITLE
[ios] Various Xcode 9 fixes

### PR DESCRIPTION
**Pull request by frederoni**
Fixes compilation for Xcode 9

- Replaced deprecated method
- Obj-C method description name is optional in Swift as of Xcode 9.
- Stricter indirect values (Color.getRed:green:blue)
 - [x] Fix backwards compatibility

@kkaefer @boundsj 👀

### CODE REVIEW

**Comment by boundsj** [comment]
Tagging in @friedbunny and @fabian-guerra

File: platform/darwin/test/MGLStyleValueTests.swift [hunk 1]
```diff
@@ -6,16 +6,34 @@ typealias MGLColor = UIColor
 #elseif os(macOS)
 typealias MGLColor = NSColor
 #endif
+
+#if swift(>=3.2)
```
Comment (friedbunny) [inline platform/darwin/test/MGLStyleValueTests.swift]: Odd they haven’t allowed `<` to work for this. 🤷‍♂️

**Comment by friedbunny** [comment]
There are still a couple failures locally for me under Xcode 9:

```
Test Case '-[MGLStyleValueTests testFunctionsWithDataDrivenProperties]' started.
/mapbox-gl-native/platform/darwin/test/MGLStyleValueTests.swift:270: error: -[MGLStyleValueTests testFunctionsWithDataDrivenProperties] : XCTAssertEqual failed: ("Optional(<MGLCompositeStyleFunction: 0x7ffbd9ddd2b0,             interpolationMode = 2,             stops = {
    0 =     {
        green = 5;
    };
    10 =     {
        green = 5;
    };
    20 =     {
        green = 20;
    };
    15 =     {
        green = 20;
    };
},             attributeName = color,             defaultValue = 2,             interpolationBase = 1.000000>)") is not equal to ("Optional(<MGLCompositeStyleFunction: 0x7ffbd9dd4cf0,             interpolationMode = 2,             stops = {
    20 = (null);
    0 =     {
        green = 5;
    };
    10 = (null);
    15 = (null);
},             attributeName = color,             defaultValue = 2,             interpolationBase = 1.000000>)") - 
```

> XCTAssertEqual failed: ("Optional(<MGLCompositeStyleFunction: 0x7ffbd9ddd2b0,             interpolationMode = 2,             stops = {
https://github.com/mapbox/mapbox-gl-native/blob/4fcdb391a82d41985b32bfa50c34c744af888be4/platform/darwin/test/MGLStyleValueTests.swift#L270

_______

```
*** Assertion failure in NSObject *MGLStyleValueTransformer<float, NSNumber *, float, NSNumber *>::toRawStyleSpecValue(MGLStyleFunction<ObjCType> *__strong) [MBGLType = float, ObjCType = NSNumber *, MBGLElement = float, ObjCEnum = NSNumber *]_block_invoke(), /mapbox-gl-native/platform/darwin/src/MGLStyleValue_Private.h:289
/mapbox-gl-native/platform/darwin/test/MGLStyleValueTests.swift:284: error: -[MGLStyleValueTests testFunctionsWithDataDrivenProperties] : failed: caught "NSInternalInconsistencyException", "Stop outputs should be MGLConstantStyleValues”
```

https://github.com/mapbox/mapbox-gl-native/blob/4fcdb391a82d41985b32bfa50c34c744af888be4/platform/darwin/test/MGLStyleValueTests.swift#L284

**Comment by friedbunny** [comment]
Let’s split 4fcdb39 and the ios-sim stuff off into its own PR(s) — will be easier to approve/merge and track in the future.

**Comment by boundsj** [comment]
https://github.com/mapbox/mapbox-gl-native/pull/10055/commits/66189856ab12075cc1aaf261468e885bd2e7a6d1 works around Swift → ObjC bridging issues that I don't yet understand. The gist is that the nested dictionaries required for composite functions don't transfer from Swift to ObjC correctly nowadays. I'm not sure when I'll have time to revisit this but I think it may be ok to merge 66189856ab12075cc1aaf261468e885bd2e7a6d1 for now to make the tests pass again.

I'm not sure about `MGLNSStringAdditionsTests` check of `XCTAssertEqualObjects([@"Improve a map" mgl_titleCasedStringWithLocale:locale], @"Improve a Map");`. Pre Xcode 9 the result of the call to `mgl_titleCasedStringWithLocale:` is @"Improve a Map". Xcode 9 and later the result is @"Improve A Map" with a capital "A". @1ec5 can you please investigate this one?

**Comment by 1ec5** [comment]
> Pre Xcode 9 the result of the call to mgl_titleCasedStringWithLocale: is @"Improve a Map". Xcode 9 and later the result is @"Improve A Map" with a capital "A".

“A” appears to be tagged as `NSLinguisticTagOtherWord`, whereas it’s more specifically tagged as `NSLinguisticTagDeterminer` on macOS and in iOS 10.

**Comment by 1ec5** [comment]
4e89282c4d1e7da438b85ddba4fd8d4eb70b727f enumerates the string’s lexical classes (parts of speech) according to the passed-in locale’s default orthography. `+[NSLocale defaultOrthographyForLanguage:]` is only available in iOS 11 and macOS 10.13, so I had to enclose it in a `+respondsToSelector:` check. Xcode 9 wants me to use `@available` instead, but the iOS build bot is currently on Xcode 8.3, so I had to suppress that warning.

**Pull request merged**